linera-core/src/client/mod.rs

```diff
@@ -1203,143 +1159,144 @@ where
             .known_committees()
             .await
             .map_err(|_| NodeError::InvalidChainInfoResponse)?;
+
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
-        let mut tracker = MultichainTracker::new(tracker);
         let mut other_sender_chains = Vec::new();
-        let remote_log = info.requested_received_log.clone();
-        let remote_node_chains_view = info.requested_received_log.into_iter().fold(
-            BTreeMap::<ChainId, Vec<BlockHeight>>::new(),
+        let remote_log = info.requested_received_log;
+        let remote_node_chains_view = remote_log.iter().fold(
+            BTreeMap::<ChainId, BlockHeight>::new(),
             |mut chain_to_info, entry| {
                 chain_to_info
                     .entry(entry.chain_id)
-                    .or_default()
-                    .push(entry.height);
+                    .and_modify(|h| *h = entry.height.max(*h))
+                    .or_insert(entry.height);
                 chain_to_info
             },
         );
 
-        // Collect all certificates that have been successfully processed.
-        let mut certificates: Vec<Certificate> = Vec::new();
-
-        for (chain_id, mut block_batch) in remote_node_chains_view {
-            block_batch.sort();
+        // Obtain the chain heights we already have from the local node.
+        let futures = remote_node_chains_view
+            .keys()
+            .map(|sender_chain_id| async move {
+                let query = ChainInfoQuery::new(*sender_chain_id);
+                let local_response = self.local_query(query).await?;
+                Ok::<_, NodeError>(local_response.info.next_block_height)
+            })
+            .collect::<Vec<_>>();
+        let local_next_heights = stream::iter(futures)
+            .buffer_unordered(chain_worker_limit)
+            .try_collect::<Vec<_>>()
+            .await?;
 
-            let Some(last_height) = block_batch.last().copied() else {
-                continue;
-            };
+        // We keep track of the height we've successfully downloaded and checked, per chain.
+        let mut downloaded_heights = BTreeMap::new();
 
-            self.advance_with_local(chain_id, &mut block_batch, &mut tracker)
-                .await?;
+        let certificate_hashes = future::try_join_all(
+            remote_node_chains_view
+                .into_iter()
+                .zip(local_next_heights)
```

Comment:
Is it guaranteed that the order of elements in `local_next_heights` will be the same after collecting it through `buffer_unordered`?

Reply:
Good catch! It isn't! (I had `join_all` here first, then changed it because of the worker limit. :facepalm:)